### ISSUE TITLE
Memory leak using nodriver with Chrome v124

### ISSUE THREAD

**Issue by Abdelrahman-Hekal** [ultrafunkamsterdam/undetected-chromedriver#1851]
Good day @ultrafunkamsterdam 
Thanks for the amazing nodriver project, I'm experiencing a continuous memory leak on a larger project when used nodriver and I tested this leak over the sample code below and as long as the code is running, it is leaking memory by 2MB per web page visited which is causing the program to be leaking around 10 GB when left running for 8 hours.

The code ran on Windows 11 and Linux where the same behaviour is noticed for both OS.

import nodriver as uc
import asyncio

async def main():

    url = "https://www.carrefouregypt.com/mafegy/en/c/NFEGY4000000?currentPage="
    browser = await uc.start(browser_args=["--incognito"])
    for i in range(1, 100):
        page, content = None, None
        page = await asyncio.wait_for(browser.get(url + str(i)), timeout=30) 
        content = await asyncio.wait_for(page.get_content(), timeout=30) 

uc.loop().run_until_complete(main())

**Comment by ilike2burnthing** [comment]
https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/1844

**Comment by Abdelrahman-Hekal** [comment]
Thanks @ilike2burnthing for referring to the uc issue but unfortunately the unclosed chrome processes of undetected-chrome is different than this issue with the nodrive package. The issue that I face is a memory leak within the python process itself as shown in the snapshot below.

![image](https://github.com/ultrafunkamsterdam/undetected-chromedriver/assets/46990368/b6a13805-9545-4f24-9295-79506b183bed)

**Comment by Abdelrahman-Hekal** [comment]
Update: I have tried three different browsers so far (Chrome, Brave and Edge) and the memory leak is still observed with all of the browsers so far

**Comment by ilike2burnthing** [comment]
There are a few other open issues pertaining to leaks, though again they seem to be for UC:
- https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/1735
- https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/1826

As a test, try reverting to Chrome v119 and v115. If you need clean installers (check the signatures, they're legitimate offline Chrome installers) - https://filecr.com/windows/google-chrome/?id=847089664000

You'll need to uninstall Chrome first, otherwise it won't install the older version if a newer version is present, and then reinstall normally once done with testing, as it won't auto-update.

**Comment by zuiqinghe2021** [comment]
I encountered the same problem, and switching to anI encountered the same problem, and switching to an still didn't resolve it?

**Comment by zuiqinghe2021** [comment]
I encountered the same issue, and switching to an older version of the browser still didn't resolve it. Are there any good solutions?

**Comment by jwxce** [comment]
Is there a solution to this problem?

**Comment by develfe** [comment]
I noticed the same problem. If I iterate over an array of links, **get_content()** increases the memory usage of the **python process** each time. (memory leak).

**Comment by life-live** [comment]
Is there a solution to this problem?

![image](https://github.com/ultrafunkamsterdam/undetected-chromedriver/assets/96289868/64a17474-3bb9-47ff-9076-c88117e23fd3)

**Comment by Abdelrahman-Hekal** [comment]
@life-live no solution is found yet so far, we highly appreciate it if you can give it some time when possible @ultrafunkamsterdam

**Comment by zxsleebu** [comment]
i think that the problem is in all functions which return elements.
to be exact memory leak happens in the moment when script gets current document to parse found nodes.
`doc: cdp.dom.Node = await self.send(cdp.dom.get_document(-1, True))`
as far as i understand garbage collector for some reason can't handle this variable and everytime the function gets called it stores more and more copies of the document in memory. probably because of the closure where child of the object points to the object.

**Comment by Abdelrahman-Hekal** [comment]
This fix really works, Great job @zxsleebu

**Comment by RandomStrangerOnTheInternet** [comment]
Works great, hoping this gets merged pretty soon!

**Comment by RandomStrangerOnTheInternet** [comment]
@ultrafunkamsterdam I've tried to create a pull request for the nodriver repository regarding the previous change however ` An owner of this repository has limited the ability to open a pull request to users that are collaborators on this repository. ` prevents me from creating a pull request with the change of @zxsleebu (https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/1851#issuecomment-2167380451)

**Comment by zxsleebu** [comment]
> @ultrafunkamsterdam I've tried to create a pull request for the nodriver repository regarding the previous change however ` An owner of this repository has limited the ability to open a pull request to users that are collaborators on this repository. ` prevents me from creating a pull request with the change of @zxsleebu (https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/1851#issuecomment-2167380451)

yeah sadly that's the issue I also encountered. that's why I posted my temp fix as a gist

**Comment by FDDQA** [comment]
Fortunately, this fix doesn't work. I also tried using a temporary file and adding a line to the library file - each time a browser request.get (url) consumes about 100 KB. Also didn't help: page.close, page del, gc.collect(), etc.

**Comment by ultrafunkamsterdam** [comment]
Thanks zxsleebu. It's implemented in 0.31

**Comment by shoaibwaqar665** [comment]
![image_2024-10-08_161457447](https://github.com/user-attachments/assets/5c47559f-cf91-4cab-9406-3a93a2d87264)


function after execution

**Comment by hkattab** [comment]
Try to introduce a timeout between driver requests, I had the same problem for a while when running the process as a task within the scheduler, so far it is working fine. I added a minimum 5000ms delay to help ensure that no overlap occurs.
FYI: the randomness is just to mimic human behavior!

randomValue = (CInt(Math.Floor((15 - 1 + 1) * Rnd())) + 1) * 1000 + 5000
Threading.Thread.Sleep(randomValue)
driver.Navigate.GoToUrl(url)

**Comment by egor-ryashin** [comment]
I have the issue in 0.39:
```
pip freeze | grep nodriver
nodriver==0.39
```

**Comment by tonynsx** [comment]
Hello

I'm having the same issue with nodriver.  I found this below.  Anyone tried this?

https://gist.github.com/zxsleebu/3602dff4be43c73e32e7d2884b7a1ac3
